README.md
# Wunsen

Wunsen transliterates/transcribes from other languages into Thai.

วุ้นเส้นสามารถทับศัพท์จากภาษาต่างๆ มาเป็นภาษาไทย

Currently support:

- Japanese (from Hepburn romanization)
- Korean (from Revised Romanization)
- Mandarin (from Hanyu Pinyin)
- Vietnamese (Latin script)

Demo [here](https://wunsen.herokuapp.com/).

## Installation

Requirement:

Python >= 3.7

[khanaa](https://github.com/cakimpei/khanaa)

```
pip install wunsen
```

## Usage

```python
from wunsen import ThapSap

# Japanese
thap_ja = ThapSap('ja')
thap_ja.thap('ohayō')
# => 'โอฮาโย'

# without macron
thap_ja_no_macron = ThapSap('ja', input='Hepburn-no diacritic')
thap_ja_no_macron.thap('ohayou')
# => 'โอฮาโย'

# another transliteration system for Japanese
thap_ja = ThapSap('ja', system='RI35')
thap_ja.thap('ohayō')
# => 'โอะฮะโย'

# Korean
thap_ko = ThapSap('ko')
thap_ko.thap('annyeonghaseyo')
# => 'อันนย็องฮาเซโย'

# Mandarin (Pinyin with tone diacritics is not supported yet.)
thap_zh = ThapSap('zh', system='RI49')
thap_zh.thap('ni3 hao3')
# => 'หนี ห่าว'

thap_zh = ThapSap('zh', system='THC43')
thap_zh.thap('ni3 hao3')
# => 'หนี เห่า'

# Vietnamese
thap_vi = ThapSap('vi')
thap_vi.thap('xin chào')
# => 'ซีน จ่าว'
```

## Transcription/Transliteration System in Wunsen

There might be some differences between Wunsen result and the intended result from the actual system, so please review the results.

- Japanese =>
    - หลักเกณฑ์การทับศัพท์ภาษาญี่ปุ่น (สำนักงานราชบัณฑิตยสภา พ.ศ. 2561) 'ORS61'
    - หลักเกณฑ์การทับศัพท์ภาษาญี่ปุ่น (ราชบัณฑิตยสถาน พ.ศ. 2535) 'RI35'
- Korean => หลักเกณฑ์การทับศัพท์ภาษาเกาหลี (ราชบัณฑิตยสถาน พ.ศ. 2555) 'RI55'
- Mandarin =>
    - หลักเกณฑ์การทับศัพท์ภาษาจีน (ราชบัณฑิตยสถาน พ.ศ. 2549) 'RI49'
    - เกณฑ์การถ่ายทอดเสียงภาษาจีนแมนดารินด้วยอักขรวิธีไทย (คณะกรรมการสืบค้นประวัติศาสตร์ไทยในเอกสารภาษาจีน พ.ศ. 2543) 'THC43'
- Vietnamese => หลักเกณฑ์การทับศัพท์ภาษาเวียดนาม (ราชบัณฑิตยสถาน พ.ศ. 2555) 'RI55'

Wunsen is not affiliated with proposers of the above transliteration systems.

วุ้นเส้นไม่มีส่วนเกี่ยวข้องกับผู้กำหนดแนวทางการทับศัพท์แต่อย่างใด

### Notes

#### Syllabification Issues

Wunsen might break syllables in incorrect place. You might have to add apostrophe:

```python
thap_ja.thap("honya | hon'ya")
# => "โฮเนีย | ฮงยะ"

thap_ko.thap("waengwaeng, maeum | waeng'waeng, ma'eum")
# => "แว็นกแว็ง, แมอุม | แว็งแว็ง, มาอึม"
```

#### Mandarin Tone Sandhi

For Mandarin, both transliteration systems specify that we should apply third tone sandhi rule to the Thai result. Wunsen will automatically apply it, but you can turn it off.

```python
thap_zh_no_sandhi = ThapSap('zh', option={'sandhi': False})
thap_zh_no_sandhi.thap('ni3 hao3')
# => 'หนี่ เห่า' / ni3 hao3

# if we turn it on
thap_zh_with_sandhi = ThapSap('zh', option={'sandhi': True})
thap_zh_with_sandhi.thap('ni3 hao3')
# => 'หนี เห่า' / ni2 hao3

thap_zh_with_sandhi.thap('ni3hao3')
# => 'หนีเห่า' / ni2hao3

# examples from wikipedia
thap_zh_with_sandhi.thap('bao3guan3 hao3')
# => 'เป๋าก๋วน เห่า' / bao2guan2 hao3

thap_zh_with_sandhi.thap('lao3 bao3guan3')
# => 'เหล่า เป๋าก่วน' / lao3 bao2guan3
```

Wunsen doesn't apply 不 (bù) and 一 (yī) tone rules as they are difficult to recognize in Pinyin.

#### Japanese long vowels

Although we should transcribe two short vowels from different origins, that are next to each other, as two short vowels (not one long vowel), Wunsen cannot cover this case reliably.

```python
thap_ja.thap("公子 kōshi | 子牛 koushi | 石井 Ishii | ただいま tadaima")
# => "公子 โคชิ | 子牛 โคอูชิ | 石井 อิชิอิ | ただいま ทาไดมะ"
# kōshi, koushi, Ishii are fine but tadaima is ta-dai-ma in Wunsen instead of ta-da-i-ma

thap_ja_no_macron.thap("公子 koushi | 子牛 koushi | 石井 Ishii | ただいま tadaima")
# => "公子 โคชิ | 子牛 โคชิ | 石井 อิชี | ただいま ทาไดมะ"
# they're transcribed as kou-shi | kou-shi | i-shii | ta-dai-ma so they're incorrect except 公子
```

#### Spacing in Thai

If we want to follow the actual transcription/transliteration system, in some cases, space between syllables or words might have to be deleted in Thai result.

For example, หลักเกณฑ์การทับศัพท์ภาษาเวียดนาม (ราชบัณฑิตยสถาน พ.ศ. 2555) (Vietnamese system) specifies that space in Vietnamese place names should be deleted, but in personal name, the space should still be there as in Vietnamese.

Because it depends on the situation, Wunsen will leave spacing as it is.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.0.3] - 2022-07-22

### Added

- New transliteration system for Japanese: หลักเกณฑ์การทับศัพท์ภาษาญี่ปุ่น (ราชบัณฑิตยสถาน พ.ศ. 2535) 'RI35'

## [0.0.2] - 2022-07-12

### Added

- Mandarin (from Hanyu Pinyin with tone numbers): หลักเกณฑ์การทับศัพท์ภาษาจีน (ราชบัณฑิตยสถาน พ.ศ. 2549) 'RI49', เกณฑ์การถ่ายทอดเสียงภาษาจีนแมนดารินด้วยอักขรวิธีไทย (คณะกรรมการสืบค้นประวัติศาสตร์ไทยในเอกสารภาษาจีน พ.ศ. 2543) 'THC43'

## [0.0.1] - 2022-06-01

### Added

- Japanese (from Hepburn romanization): หลักเกณฑ์การทับศัพท์ภาษาญี่ปุ่น (สำนักงานราชบัณฑิตยสภา พ.ศ. 2561) 'ORS61'
- Korean (from Revised Romanization): หลักเกณฑ์การทับศัพท์ภาษาเกาหลี (ราชบัณฑิตยสถาน พ.ศ. 2555) 'RI55'
- Vietnamese (Latin script): หลักเกณฑ์การทับศัพท์ภาษาเวียดนาม (ราชบัณฑิตยสถาน พ.ศ. 2555) 'RI55'

src/wunsen/mandarin/zh_thapsap.py
import re
from typing import Any, Dict, List

from khanaa import SpellWord

from wunsen.splitutils.splitter import NotInDict
from wunsen.splitutils.exception import NotAvailableSystem
from .zh_splitter import SplitZh
from .zh_mapping import RI49, THC43
from .zh_sandhi import AlterThird, SPLIT_EXP

class ZhRi49:

    zh_dict = RI49
    short_forth_list = ['ai', 'an', 'ang', 'ao', 'uai', 'uang']
    aao_aang = {
        'w': ['u'],
        'y': ['i']
    }

    def __init__(self) -> None:
        pass

    def thap_kham(self, word: str) -> str:
        try:
            splitter = SplitZh(self.zh_dict)
            splitter.sys_dict = self.zh_dict
            self.split = splitter.split_syl(word)
        except NotInDict:
            return word
        result_list = []
        self.pref = {}
        spell = SpellWord(**{
            'clear_vowel': False,
            'obvious_h_low_single': False})
        for index, syl in enumerate(self.split):
            if self.is_erhua(syl):
                if len(self.split) == 1:
                    return word
                else:
                    continue
            self.is_erhua_next(index)
            self.shorten = False
            pattern = self.turn_thai(syl)
            self.select_pref(syl)
            spell.option.update(**self.pref)
            result_list.append(spell.spell_out(**pattern))

        result = ''.join(result_list)

        return result

    @staticmethod
    def is_erhua(syl: List) -> bool:
        return syl[0] == 'r' and syl[1] == ''
    
    def is_erhua_next(self, index) -> None:
        self.erhua_next = False
        if (index < len(self.split) - 1
                and self.is_erhua(self.split[index+1])):
            self.erhua_next = True

    def turn_thai(self, syl: List) -> Dict[str, Any]:
        self.select_rime(syl)
        self.select_onset(syl)
        self.select_tone(syl)
        self.short_fifth(syl)
        pattern = {
            'onset': self.onset,
            'vowel': self.rime[1],
            'silent_before': self.rime[2],
            'coda': self.rime[3],
            'tone': self.tone
        }
        if self.erhua_next:
            pattern = self.add_erhua(pattern)
        return pattern

    def select_rime(self, syl: List) -> None:
        self.rime = self.zh_dict['rime'][syl[1]]

        self.short_fourth_a(syl)

        self.rime_dict = 'rime'
        if (syl[1] in self.zh_dict['cond_rime']
                and syl[0] in self.zh_dict['cond_rime'][syl[1]]):
            self.rime = self.zh_dict['rime_cond'][syl[1]]
            self.rime_dict = 'rime_cond'

    def short_fourth_a(self, syl: List) -> None:
        if (syl[1] in self.short_forth_list
                and syl[2] == 4):
            self.shorten = True

    def select_onset(self, syl: List) -> None:
        self.onset = self.zh_dict['onset'][syl[0]]
        vowel_onset = self.zh_dict[self.rime_dict][syl[1]][0]

        if (syl[0] in self.aao_aang
                and syl[1] in self.aao_aang[syl[0]]):
            self.onset = 'อ'
        if vowel_onset != 'อ':
            if syl[0] == '':
                self.onset = vowel_onset
            else:
                self.onset = ''.join([self.onset, vowel_onset])

    def select_tone(self, syl: List) -> None:
        self.tone = self.zh_dict['tone'][syl[2]]

    def short_fifth(self, syl: List) -> None:
        if (syl[2] in [0, 5]
                and (syl[1] == 'a'
                or (syl[1] == 'i'
                and syl[0] in self.zh_dict['cond_rime'][syl[1]]))):
            self.shorten = True
            self.tone = 1

    @staticmethod
    def add_erhua(pattern: Dict[str, Any]) -> Dict[str, Any]:
        return pattern

    def select_pref(self, syl: List) -> None:
        if self.shorten == True:
            self.pref.update({'vowel_length': 'short'})
        else:
            self.pref.update({'vowel_length': 'input'})
        if (self.rime[4]
                or (self.onset == 'ฮว' and self.tone in [1, 4])):
            self.pref.update({'onset_style': 'phinthu'})
        else:
            self.pref.update({'onset_style': 'plain'})

class ZhThc43(ZhRi49):
    
    zh_dict = THC43
    short_forth_list = ['an']
    aao_aang = {
        'w': ['u'],
        'y': ['i', 'in', 'ing', 'u', 'un', 'v', 'vn']
    }

    def short_fifth(self, syl: List) -> None:
        pass

    @staticmethod
    def add_erhua(pattern: Dict[str, Any]) -> Dict[str, Any]:
        if pattern['silent_before']:
            new_silent_before = ''.join([
                pattern['silent_before'], 'ร'])
            pattern['silent_before'] = new_silent_before
        else:
            pattern.update({'silent_after': 'ร'})
        return pattern

    def select_pref(self, syl: List) -> None:
        super().select_pref(syl)
        if syl[0] == 'w' and syl[1] == 'an':
            self.pref.update({'vowel_length': 'short'})
        elif syl[1] == 'ei' and not self.erhua_next:
            self.pref.update({'silent_before_style': 'plain'})

class ThapZh:
    def __init__(
        self, system: str = 'THC43',
        input: str = 'Pinyin-number',
        option: Dict[str, Any] = {'sandhi': True}) -> None:
        """Setting
        
        :param system: Select thapsap system.
            - 'RI49' for the Royal Institute (2006/2549) system
            - 'THC43' for เกณฑ์การถ่ายทอดเสียงภาษาจีนแมนดาริน
            ด้วยอักขรวิธีไทย (2000/2543)

        :param input: Select input type.
            - 'Pinyin-number' for Hanyu Pinyin with tone number
            (ex. xie4xie5)
        
        :param option: Select option
            - for third tone sandhi
                - {'sandhi': True} for automatic third tone sandhi
                (ex. ni3 hao3 => หนีห่าว (ni2 hao3))
                - {'sandhi': False}
                (ex. ni3 hao3 => หนี่ห่าว (ni3 hao3))
        """
        if system == 'RI49' and input == 'Pinyin-number':
            self.transcript = ZhRi49()
            self.sys_dict = RI49
        elif system == 'THC43' and input == 'Pinyin-number':
            self.transcript = ZhThc43()
            self.sys_dict = THC43
        else:
            raise NotAvailableSystem
        self.option = option

    def thap(self, text: str) -> str:
        if self.option['sandhi']:
            alter_third = AlterThird(self.sys_dict)
            text = alter_third.third_alter(text)
        def replace(match):
            return self.transcript.thap_kham(match.group())
        return re.sub(SPLIT_EXP, replace, text)